fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/DisplayWindow.java

```diff
@@ -0,0 +1,531 @@
+        glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
+        if (mcVersion != null) {
+            // this emulates what we would get without early progress window
+            // as vanilla never sets these, so GLFW uses the first window title
+            // set them explicitly to avoid it using "FML early loading progress" as the class
+            String vanillaWindowTitle = "Minecraft* " + mcVersion;
+            glfwWindowHintString(GLFW_X11_CLASS_NAME, vanillaWindowTitle);
+            glfwWindowHintString(GLFW_X11_INSTANCE_NAME, vanillaWindowTitle);
+        }
+
+        long primaryMonitor = glfwGetPrimaryMonitor();
+        if (primaryMonitor == 0) {
+            LOGGER.error("Failed to find a primary monitor - this means LWJGL isn't working properly");
+            crashElegantly("Failed to locate a primary monitor.\nglfwGetPrimaryMonitor failed.\n");
+            throw new IllegalStateException("Can't find a primary monitor");
+        }
+        GLFWVidMode vidmode = glfwGetVideoMode(primaryMonitor);
+
+        if (vidmode == null) {
+            LOGGER.error("Failed to get the current display video mode.");
+            crashElegantly("Failed to get current display resolution.\nglfwGetVideoMode failed.\n");
+            throw new IllegalStateException("Can't get a resolution");
+        }
+        long window;
+        int versidx= 0;
+        final String[] lastGLError=new String[versions.length];
+        do {
+            LOGGER.info("Trying GL version "+versions[versidx][0]+"."+versions[versidx][1]);
+            glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, versions[versidx][0]); // we try our versions one at a time
+            glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, versions[versidx][1]);
+            glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+            glfwWindowHint(GLFW.GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
+            window = glfwCreateWindow(winWidth, winHeight, "Minecraft: Forge Loading...", 0L, 0L);
+            var erridx = versidx;
+            handleLastGLFWError((error, description) -> lastGLError[erridx] = String.format("Trying %d.%d: GLFW error: [0x%X]%s", versions[erridx][0], versions[erridx][1], error, description));
+            if (lastGLError[versidx] != null) {
+                LOGGER.trace(lastGLError[versidx]);
+            }
+            versidx++;
+            if (versidx== versions.length) {
+                LOGGER.error("Failed to find any valid GLFW profile. "+lastGLError[0]);
+                crashElegantly("Failed to find a valid GLFW profile.\nWe tried "+String.join(",",glVersions)+" but none of them worked.\n"+String.join("\n", Arrays.asList(lastGLError)));
+                throw new IllegalStateException("Failed to create a GLFW window with any profile");
+            }
+        } while (window == 0);
+        this.glVersion = versions[versidx-1][0]+"."+versions[versidx-1][1];
+        LOGGER.info("Loaded GL version "+this.glVersion);
```

Comment:
This should be based on the OpenGL version reported by the context and not what was asked for. As noted in my other comment, the drivers are free to return a different version and some do.

Reply:
Ok.

Reply:
Returning the value from the window now. Thanks